link/uprobe_multi.go

```diff
@@ -0,0 +1,199 @@
+package link
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/asm"
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/sys"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+// UprobeMultiOptions defines additional parameters that will be used
+// when opening a UprobeMulti Link.
+type UprobeMultiOptions struct {
+	// Path of binary to attach to.
+	Path string
+
+	// Symbol addresses. If set, overrides the address eventually parsed
+	// from the executable.
+	// Mutually exclusive with symbols UprobeMulti argument.
+	Addresses []uint64
+
+	// Offsets into functions provided by symbols array in UprobeMulti
+	// For example to set uprobes to main+5 and _start+10 call UprobeMulti
+	// with:
+	//     symbols: "main", "_start"
+	//     opt.Offset: 5, 10
+	Offsets []uint64
+
+	// Optional, array of associated ref counter offsets.
+	RefCtrOffsets []uint64
+
+	// Optional, array of associated BPF cookies.
+	Cookies []uint64
+}
+
+// addressMulti calculates the array of addresses from symbols array in the executable.
+//
+// opts must not be nil.
+func (ex *Executable) uprobeMultiAddresses(symbols []string, opts *UprobeMultiOptions) ([]uint64, error) {
+	var addresses []uint64
+
+	offsets := len(opts.Offsets)
+	for idx, symbol := range symbols {
+		address, err := ex.address(symbol)
+		if err != nil {
+			return []uint64{0}, err
+		}
+		if offsets != 0 {
+			address += opts.Offsets[idx]
+		}
+		addresses = append(addresses, address)
+	}
+	return addresses, nil
+}
+
+func (ex *Executable) UprobeMulti(symbols []string, prog *ebpf.Program, opts *UprobeMultiOptions) (Link, error) {
+	if opts == nil {
+		opts = &UprobeMultiOptions{}
+	}
+
+	syms := uint32(len(symbols))
+	addrs := uint32(len(opts.Addresses))
+	offsets := uint32(len(opts.Offsets))
+
+	if addrs != 0 && (offsets != 0 || syms != 0) {
+		return nil, fmt.Errorf("Address and Offsets/symbols are mutually exclusive: %w", errInvalidInput)
+	}
+	if offsets != 0 && offsets != syms {
+		return nil, fmt.Errorf("Offsets must be either zero or exactly symbols in length: %w", errInvalidInput)
+	}
+
+	var err error
+	if syms != 0 {
+		// Translate symbols to offsets
+		opts.Addresses, err = ex.uprobeMultiAddresses(symbols, opts)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return ex.uprobeMulti(prog, opts, 0)
+}
+
+func (ex *Executable) UretprobeMulti(symbols []string, prog *ebpf.Program, opts *UprobeMultiOptions) (Link, error) {
```

Comment:
This method completely ignores `symbols`? Also, is it even valid to specify offsets for Uretprobe?

Reply:
yes, it can have same setup as entry probe, so it needs to do same resolving.. I fixed that in current version